notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb

```diff
@@ -0,0 +1,397 @@
+    "                and agent != last_speaker and agent != self.previous_speaker\n",
+    "            ]\n",
+    "\n",
+    "        self.previous_speaker = last_speaker\n",
+    "\n",
+    "        if possible_next_speakers:\n",
+    "            next_speaker = random.choice(possible_next_speakers)\n",
+    "            return next_speaker\n",
+    "        else:\n",
+    "            return None\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Termination message detection\n",
+    "def is_termination_msg(content) -> bool:\n",
+    "    have_content = content.get(\"content\", None) is not None\n",
+    "    if have_content and \"TERMINATE\" in content[\"content\"]:\n",
+    "        return True\n",
+    "    return False"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\u001b[33mB2\u001b[0m (to chat_manager):\n",
+      "\n",
+      "Find the product of x and y, the other agents know x and y.\n",
+      "\n",
+      "--------------------------------------------------------------------------------\n",
+      "[autogen.oai.completion: 10-31 09:38:16] {788} WARNING - Completion was provided with a config_list, but the list was empty. Adopting default OpenAI behavior, which reads from the 'model' parameter instead.\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "WARNING:autogen.oai.completion:Completion was provided with a config_list, but the list was empty. Adopting default OpenAI behavior, which reads from the 'model' parameter instead.\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\u001b[33mB1\u001b[0m (to chat_manager):\n",
+      "\n",
+      "NEXT: A1\n",
+      "Do you or your team members, A2 and A3, know the values for x and y? B2 needs them to find their product.\n",
+      "\n",
+      "--------------------------------------------------------------------------------\n",
+      "[autogen.oai.completion: 10-31 09:38:19] {788} WARNING - Completion was provided with a config_list, but the list was empty. Adopting default OpenAI behavior, which reads from the 'model' parameter instead.\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "WARNING:autogen.oai.completion:Completion was provided with a config_list, but the list was empty. Adopting default OpenAI behavior, which reads from the 'model' parameter instead.\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\u001b[33mA1\u001b[0m (to chat_manager):\n",
+      "\n",
+      "Yes, A2 has the value of x and A3 knows the value of y. Let me get those for you. A2, what is the value of x? And A3, what is the value of y?\n",
```

Comment:
This is the last message. Shouldn't there be more messages in the conversation?

Reply:
I'm not sure what happened there, but anyways, I have cleared all outputs and rerun the entire script.



B2 (to chat_manager):

Find the product of x and y, the other agents know x and y.

--------------------------------------------------------------------------------
B1 (to chat_manager):

NEXT: A1
Can you or your team member provide the value of x and y?

--------------------------------------------------------------------------------
Extracted suggested_next = A1
Can you or your team member provide the value of x and y?
A1 (to chat_manager):

As A1, I don't have the values. A2, A3, do you know the values of x and y?

--------------------------------------------------------------------------------
A2 (to chat_manager):

As A2, I can confirm that the value of x is 9. However, I do not know the value of y, maybe another team member could provide it?

--------------------------------------------------------------------------------
A3 (to chat_manager):

As A3, I can confirm that the value of y is 5.

--------------------------------------------------------------------------------
A1 (to chat_manager):

Considering that the value of x is 9 and the value of y is 5, the product of x and y is 45.

--------------------------------------------------------------------------------
B1 (to chat_manager):

NEXT: B2
The product of x and y is 45.

--------------------------------------------------------------------------------
Extracted suggested_next = B2
The product of x and y is 45.
B2 (to chat_manager):

The product of x and y is 45. 
TERMINATE

--------------------------------------------------------------------------------